File: libraries/types/sprite.lua
---@diagnostic disable: return-type-mismatch, missing-return-value
local engine = require('libraries.engine')
local image = require('libraries.image')

monoe = monoe or {}

---@class monoe.sprite
---@field uid integer
monoe.sprite = {}
monoe.sprite.__index = monoe.sprite

local base = 'monoe.exe.Core.Bridge.Types.Sprite'

---Crate a new sprite. If a path is provided, it'll load it as an image
---@param path string|nil
---@return monoe.sprite
function monoe.sprite.new(path)
  local uid = engine.import(base)

  if uid == -1 then
    error('Got bad UID when creating monoe.sprite object!')
  end

  if path then
    engine.call(uid, 'LoadImage', path)
  end

  return setmetatable({ uid = uid }, monoe.sprite)
end

function monoe.sprite:clear()
  engine.call(self.uid, 'Clear')
end

function monoe.sprite:load(path)
  engine.call(self.uid, 'LoadImage', path)
end

---places the sprite at x;y
---@param x number
---@param y number
---@return number
---@return number
function monoe.sprite:position(x, y)
  if x and y then
    engine.call(self.uid, 'SetPosition', x, y)
    return x, y
  end

  return engine.call(self.uid, 'GetPosition')
end

---@param x number
---@param y number
---@return number
---@return number
function monoe.sprite:move(x, y)
  return engine.call(self.uid, 'Move', x, y)
end

---@param x number
---@param y number
---@return number
---@return number
function monoe.sprite:scale(x, y)
  return engine.call(self.uid, "Scale", x, y)
end

---@return integer
function monoe.sprite:image()
  return image.new(engine.call(self.uid, 'GetImageUID'))
end

function monoe.sprite:free()
  engine.call(self.uid, 'Free')
end

_G.monoe = monoe
_G.monoe.sprite = monoe.sprite
return monoe.sprite

File: Core/Export/Types/Sprite.cs
using Godot;
using monoe.exe.Core.Bridge;

namespace monoe.exe.Core.Export.Types;

public class Sprite : Exposable
{
  private readonly Sprite2D sprite = new();

  public void LoadImage(string path)
  {
    sprite.Texture = ImageTexture.CreateFromImage(Image.LoadFromFile(path));
  }

  public void Clear()
  {
    sprite.Texture = null;
  }

  public void SetPosition(double x, double y)
  {
    sprite.Position = new((float)x, (float)y);
  }

  public object[] GetPosition()
  {
    return [sprite.Position.X, sprite.Position.Y];
  }

  public object[] Move(double x, double y)
  {
    sprite.Position = new(sprite.Position.X + (float)x, sprite.Position.Y + (float)y);
    return [sprite.Position.X, sprite.Position.Y];
  }

  public object[] Scale(double x, double y)
  {
    sprite.Scale = new((float)x, (float)y);
    return [sprite.Scale.X, sprite.Scale.Y];
  }

  protected override void _Free()
  {
    sprite.QueueFree();
  }

  public override Node NRef() => sprite;
}